File: scripts/search_chunks.py
#!/usr/bin/env python3
"""
Semantic search over pgvector chunks using OpenAI embeddings.

Usage:
    python3 scripts/search_chunks.py "your query here"
    python3 scripts/search_chunks.py "query" --top_k 10
    python3 scripts/search_chunks.py "query" --source_file incidents
    python3 scripts/search_chunks.py "query" --type adr
"""

import argparse
import json
import os
import sys
from pathlib import Path

from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parent.parent / '.env')

import openai
import psycopg2

EMBEDDING_MODEL = 'text-embedding-3-small'

DB_CONFIG = {
    'host': os.getenv('POSTGRES_HOST', 'localhost'),
    'port': int(os.getenv('POSTGRES_PORT', '5432')),
    'dbname': os.getenv('POSTGRES_DB', 'proshop'),
    'user': os.getenv('POSTGRES_USER', 'proshop'),
    'password': os.getenv('POSTGRES_PASSWORD', 'proshop'),
}


def embed_query(text, client):
    """Embed a single query string."""
    response = client.embeddings.create(model=EMBEDDING_MODEL, input=[text])
    return response.data[0].embedding


def search(query, top_k=5, source_file=None, file_type=None):
    """
    Semantic search over chunks.

    Args:
        query: Search query string.
        top_k: Number of results to return.
        source_file: Filter by source_file (substring match).
        file_type: Filter by file_path directory/type (substring match).

    Returns:
        List of dicts with text, score, and metadata.
    """
    api_key = os.getenv('OPENAI_API_KEY')
    if not api_key:
        print('ERROR: OPENAI_API_KEY not set.')
        sys.exit(1)

    client = openai.OpenAI(api_key=api_key)
    embedding = embed_query(query, client)
    embedding_str = '[' + ','.join(str(x) for x in embedding) + ']'

    # Build query with optional filters
    conditions = []
    params = []

    if source_file:
        conditions.append('source_file ILIKE %s')
        params.append(f'%{source_file}%')

    if file_type:
        conditions.append('file_path ILIKE %s')
        params.append(f'%{file_type}%')

    where_clause = ''
    if conditions:
        where_clause = 'WHERE ' + ' AND '.join(conditions)

    sql = f'''
        SELECT
            text,
            source_file,
            file_path,
            title,
            parent_headings,
            keywords,
            summary,
            1 - (embedding <=> %s::vector) AS similarity
        FROM chunks
        {where_clause}
        ORDER BY embedding <=> %s::vector
        LIMIT %s
    '''
    params = [embedding_str] + params + [embedding_str, top_k]

    # Fix: params order for WHERE between the two embedding refs
    # Rebuild properly
    if conditions:
        sql = f'''
            SELECT
                text,
                source_file,
                file_path,
                title,
                parent_headings,
                keywords,
                summary,
                1 - (embedding <=> %s::vector) AS similarity
            FROM chunks
            {where_clause}
            ORDER BY embedding <=> %s::vector
            LIMIT %s
        '''
        params = [embedding_str] + [f'%{source_file}%'] * (1 if source_file else 0) + \
                 [f'%{file_type}%'] * (1 if file_type else 0) + [embedding_str, top_k]
    else:
        sql = '''
            SELECT
                text,
                source_file,
                file_path,
                title,
                parent_headings,
                keywords,
                summary,
                1 - (embedding <=> %s::vector) AS similarity
            FROM chunks
            ORDER BY embedding <=> %s::vector
            LIMIT %s
        '''
        params = [embedding_str, embedding_str, top_k]

    conn = psycopg2.connect(**DB_CONFIG)
    with conn.cursor() as cur:
        cur.execute(sql, params)
        rows = cur.fetchall()
    conn.close()

    results = []
    for row in rows:
        results.append({
            'text': row[0][:300] + ('...' if len(row[0]) > 300 else ''),
            'full_text': row[0],
            'source_file': row[1],
            'file_path': row[2],
            'title': row[3],
            'parent_headings': row[4],
            'keywords': row[5],
            'summary': row[6],
            'similarity': round(float(row[7]), 4),
        })

    return results


def main():
    parser = argparse.ArgumentParser(description='Semantic search over project chunks')
    parser.add_argument('query', help='Search query')
    parser.add_argument('--top_k', type=int, default=5, help='Number of results (default: 5)')
    parser.add_argument('--source_file', type=str, default=None, help='Filter by source_file substring')
    parser.add_argument('--type', type=str, default=None, dest='file_type', help='Filter by file_path substring (e.g. incidents, adr)')
    args = parser.parse_args()

    results = search(args.query, top_k=args.top_k, source_file=args.source_file, file_type=args.file_type)

    print(f'\n{"="*60}')
    print(f'Query: "{args.query}"')
    if args.source_file:
        print(f'Filter: source_file ~ "{args.source_file}"')
    if args.file_type:
        print(f'Filter: file_path ~ "{args.file_type}"')
    print(f'Results: {len(results)}')
    print(f'{"="*60}\n')

    for i, r in enumerate(results, 1):
        print(f'--- Result {i} (score: {r["similarity"]}) ---')
        print(f'  Source: {r["source_file"]}')
        print(f'  Path:   {r["file_path"]}')
        print(f'  Title:  {r["title"]}')
        if r["parent_headings"]:
            print(f'  Section: {" > ".join(r["parent_headings"])}')
        print(f'  Text:   {r["text"][:200]}...')
        print()


if __name__ == '__main__':
    main()

File: scripts/search_chunks.js
#!/usr/bin/env node
/**
 * Semantic search over pgvector chunks using OpenAI embeddings.
 *
 * Usage:
 *   node scripts/search_chunks.js "your query here"
 *   node scripts/search_chunks.js "query" --top_k 10
 *   node scripts/search_chunks.js "query" --source_file incidents
 *   node scripts/search_chunks.js "query" --type adr
 */

import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { parseArgs } from 'util';
import dotenv from 'dotenv';
import OpenAI from 'openai';
import pg from 'pg';

const __dirname = dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: resolve(__dirname, '..', '.env') });

const EMBEDDING_MODEL = 'text-embedding-3-small';

const DB_CONFIG = {
    host: process.env.POSTGRES_HOST || 'localhost',
    port: parseInt(process.env.POSTGRES_PORT || '5432', 10),
    database: process.env.POSTGRES_DB || 'proshop',
    user: process.env.POSTGRES_USER || 'proshop',
    password: process.env.POSTGRES_PASSWORD || 'proshop',
};

async function embedQuery(text, openai) {
    const response = await openai.embeddings.create({
        model: EMBEDDING_MODEL,
        input: [text],
    });
    return response.data[0].embedding;
}

export async function search(query, { topK = 5, sourceFile = null, fileType = null } = {}) {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
        console.error('ERROR: OPENAI_API_KEY not set.');
        process.exit(1);
    }

    const openai = new OpenAI({ apiKey });
    const embedding = await embedQuery(query, openai);
    const embeddingStr = '[' + embedding.join(',') + ']';

    const conditions = [];
    const params = [embeddingStr];
    let paramIdx = 2;

    if (sourceFile) {
        conditions.push(`source_file ILIKE $${paramIdx++}`);
        params.push(`%${sourceFile}%`);
    }

    if (fileType) {
        conditions.push(`file_path ILIKE $${paramIdx++}`);
        params.push(`%${fileType}%`);
    }

    const whereClause = conditions.length ? 'WHERE ' + conditions.join(' AND ') : '';

    // $1 is embedding, filters are $2+, last two are embedding again + limit
    params.push(embeddingStr);
    const embIdx2 = paramIdx++;
    params.push(topK);
    const limitIdx = paramIdx++;

    const sql = `
        SELECT
            text,
            source_file,
            file_path,
            title,
            parent_headings,
            keywords,
            summary,
            1 - (embedding <=> $1::vector) AS similarity
        FROM chunks
        ${whereClause}
        ORDER BY embedding <=> $${embIdx2}::vector
        LIMIT $${limitIdx}
    `;

    const client = new pg.Client(DB_CONFIG);
    await client.connect();
    const result = await client.query(sql, params);
    await client.end();

    return result.rows.map((row) => ({
        text: row.text.length > 300 ? row.text.slice(0, 300) + '...' : row.text,
        full_text: row.text,
        source_file: row.source_file,
        file_path: row.file_path,
        title: row.title,
        parent_headings: row.parent_headings,
        keywords: row.keywords,
        summary: row.summary,
        similarity: parseFloat(parseFloat(row.similarity).toFixed(4)),
    }));
}

async function main() {
    const { values, positionals } = parseArgs({
        allowPositionals: true,
        options: {
            top_k: { type: 'string', default: '5' },
            source_file: { type: 'string', default: '' },
            type: { type: 'string', default: '' },
        },
    });

    const query = positionals[0];
    if (!query) {
        console.error('Usage: node search_chunks.js "query" [--top_k N] [--source_file X] [--type Y]');
        process.exit(1);
    }

    const topK = parseInt(values.top_k, 10);
    const sourceFile = values.source_file || null;
    const fileType = values.type || null;

    const results = await search(query, { topK, sourceFile, fileType });

    console.log('\n' + '='.repeat(60));
    console.log(`Query: "${query}"`);
    if (sourceFile) console.log(`Filter: source_file ~ "${sourceFile}"`);
    if (fileType) console.log(`Filter: file_path ~ "${fileType}"`);
    console.log(`Results: ${results.length}`);
    console.log('='.repeat(60) + '\n');

    for (let i = 0; i < results.length; i++) {
        const r = results[i];
        console.log(`--- Result ${i + 1} (score: ${r.similarity}) ---`);
        console.log(`  Source: ${r.source_file}`);
        console.log(`  Path:   ${r.file_path}`);
        console.log(`  Title:  ${r.title}`);
        if (r.parent_headings && r.parent_headings.length) {
            console.log(`  Section: ${r.parent_headings.join(' > ')}`);
        }
        console.log(`  Text:   ${r.text.slice(0, 200)}...`);
        console.log();
    }
}

main();
